### ISSUE TITLE
manual setup for kubernetes

### ISSUE THREAD

**Issue by sunyi000** [abcdesktopio/docs#7]
Hi @alexandredevely ,

thanks so much for releasing v3.0

I started experimenting the release 3.0 by following the docs.

I just want to report a minor thing here. 

In the "Manual installation step by step" part of the "Setup abcdesktop for kubernetes" section,
running `kubectl create -f abcdesktop-3.0.yaml` did not create the abcdesktop-config configmap. (probably I missed something..)

but after I manually download od.config and `kubectl create configmap.... `solves the problem, and then everything started to work.

the install script install3.0.sh worked.

thanks very much, I will come back with more questions as I move along in my experiments

Yi

**Comment by alexandredevely** [comment]
Hello Yi Sun,

That you for your feedback, I've fixed the installation process, and add step 4 for the configmap.
[kubernetes_abcdesktop](https://www.abcdesktop.io/3.0/setup/kubernetes_abcdesktop)

You are an you are amazing and so fast !
I've just git push yerterday the new release, and i would like to run some tests, before sending you the update. You are so fast !

I think to reduce image application size. It take up to 4 GB for libreoffice with all languages support build in, based on ubuntu 22.04.
May be alpine could be a good choice ?
Today only application xterm.d use alpine as source. 

The uidNumber and gidNumber is supported, and user create file with the correct posix uid/gid. All container process start with the uid/gid.

The next 3.1 release will support the nvidia/cuda for application.

See you, and thank you again. You've got that eagle eye.

**Comment by sunyi000** [comment]
Hi @alexandredevely 

Thank you for fixing it.

I was doing some other works in the past couple of weeks and just finished yesterday, so I thought I could revisit abcdesktop, then I gladly discovered lots of updates in both github and the docs website :)

Regarding the application images, it's actually what I want to ask next...:)

I think Alpine could help reducing the size but in some applications Ubuntu maybe still needed because it has a lot of utilities built-in.

My questions regarding apps:

What would be the standard steps to build our own app in v3.0? I see in 3.0,to build an app, it requires a json file(looks like an output of docker inspect) to be posted to the API endpoint. 
In v1.1 I could just docker build from local dockerfile, and abcdedktop just automatically detects the image.

how do we make local docker image available in abcdesktop3.0, without using docker registery.

I tried to build a local docker image but the abcdesktop does not pick up the new image automatically. Then I did docker inspect to output the json and post it to the endpoint. Now it seems abcdesktop trying to look for the image from dockerhub.  

If we build our own docker image from dockerfile , which base image we should use? for v1.1 I build all our own apps with oc.template.gtk , is this still the same?

Thanks again for all your help
Yi

**Comment by alexandredevely** [comment]
Hi @sunyi000,

Thanks for your question.

The release requires a json file(looks like an output of docker inspect) to be posted to the API endpoint.
Without using docker registry, if you run a 

```
kubectl get pods -n abcdesktop
```

You should read a pod named  `pull-XXXXXXXX` in `ErrImagePull`

```
pull-alpine-2048-dev-desktop-4aeaffbe2eccfcc44f625f179d0d   0/1     ErrImagePull       0                 14s
```

You can delete it 

```
kubectl delete pods pull-alpine-2048-dev-desktop-4aeaffbe2eccfcc44f625f179d0d -n abcdesktop
pod "pull-alpine-2048-dev-desktop-4aeaffbe2eccfcc44f625f179d0d" deleted
```

Your application is already present on your system, if you build it localy. So you don't need a private registry to build.
( I will write an update to disable pulling )

About images for release 3.0.
All images exist with the tag 3.0, for example 

```
abcdesktopio/oc.template.gtk:3.0 
abcdesktopio/oc.template:3.0
```

So if you build our own docker image from dockerfile , just replace `1.1` with `3.0`.

I think it's time to reduce the size of images. I agree with you, too keep ubuntu distrib.
Thank you again.

Alexandre

**Comment by sunyi000** [comment]
thanks @alexandredevely 

1. I got my local app appear in the abcdesktop, although it's not fully functioning yet, I think i can fix it. thank you

other questions:
1. I got ldap login working..however the home directory looks a bit weird.  
I have all users' home dir on nfs drive at /nfs/home, and I created a soft link to /home . I specified /nfs/home in od.config as the value of desktop.hostPathRoot . homedirectoryPath is hostPath. 

user: yisun is provided by ldap via sssd, the ldap attribute homeDirectory is /home/yisun . gidNumber is 1102(users). uid is yisun. I have some files in the directory.

After redeploy it, my home directory is still /home/balloon, but the ownership of /home/balloon yisun:yisun 
I was expecting my home to be /home/yisun, and ownership of yisun:users.

Did I miss anything in the config?


2. is the "User data persistence" section in "How-to" document still valid? I see in the od.config, persistentvolume is not supported in v3.0.

3. if I have another nfs drive, not home directory, for example, /nfs/g/data with some folders and files owned by the same users in the same ldap. To make it available in the desktop, should I just append the volume in the pyos section in abcdesktop.yaml?

thanks for all the help..
Yi

**Comment by alexandredevely** [comment]
Hello @sunyi000,

"After redeploy it, my home directory is still /home/balloon, but the ownership of /home/balloon yisun:yisun
I was expecting my home to be /home/yisun, and ownership of yisun:users."
That's correct. the home dir is /home/balloon ( but this is just the name of this directory ) 
( Ok i understand that you except  to have a homeDirectory like /home/yisun ) ( i check if i can change balloon name with the ldap values )
But uidNumber and guiNumber should be correct and read from your ldap directory.

2.  persistentvolume is not supported in v3.0. ( for ephemeral containers). The "How-to"  is only  for release 1.0 ( fixed )

3. I take this point. It works with CIFS, so i mean create an update for NFS should work. But you will need to define a tag to ask pyos to add this volume. ( rule can be a dummy True ( always True ) or ldap memberof, or ip network.... )

5. File Manager seems missing in the 3.0. Yes and no. FileManager is now an application. So you need to add it manually ( for security )

6. Yes there is a cache in mongodb. Why ?
Pyos does not need and can not reach containerd  ( or dockerd ). So if you remove an application then pyos can't get the event. So we need to use the REST API 

I ve updated the page[https://www.abcdesktop.io/3.0/setup/kubernetes_abcdesktop_applications/]( https://www.abcdesktop.io/3.0/setup/kubernetes_abcdesktop_applications/)

REST API `methods` description for `/API/manager/image`

| Method   | Type  |                     
|----------|----------|
| `GET`    | http request list images in mongo db `image` collection  |
| `PUT`    | http request update or insert images in mongo db `image` collection, then create a pull pod to fetch images | 
| `POST`   | http request update or insert images in mongo db `image` collection. This method does not pull images.   | 
| `DELETE` | http request delete images in mongo db `image` collection    |

| Method   | Sample | 
|----------|--------|
| `GET`    | `curl -X GET -H 'Content-Type: text/javascript' http://localhost:30443/API/manager/image`                                                      |
| `PUT`    | `curl -X PUT -H 'Content-Type: text/javascript' http://localhost:30443/API/manager/image -d@xterm.d.json`  |
| `POST`   | `curl -X POST -H 'Content-Type: text/javascript' http://localhost:30443/API/manager/image -d@xterm.d.json`      |
| `DELETE` | `curl -X DELETE -H 'Content-Type: text/javascript' http://localhost:30443/API/manager/image/efbb56e0c579d1945fd8f4a4d955e08d7801208c953e03fe6d4d274edd1904c9`                                                    |


See you
Alexandre

**Comment by sunyi000** [comment]
thank you @alexandredevely 

for the gid bit on /home/balloon, 

inside the desktop
if I do a `id -g yisun`, I can see the gid is 1102 which is correct. 
if I do a `cat /etc/group`, I see the last line being `yisun:x:1102`, which has the correct gid but wrong name.

outside the desktop, on the k8s node,
`id -g yisun` shows the correct number 1102. 
`ls /home/` shows the ownership of /home/yisun is `yisun:users`
`cat /etc/group`, does not have yisun:x:1102 at the last line

so the difference is the last line of /etc/group. I checked the od.config. 
I see in oc.app /etc/group is linked to a mounted directory /var/secret/abcdesktop/localaccount , and it seems related to configmap 'localaccount', but I can't find the configmap :)

thanks

**Comment by alexandredevely** [comment]
Hello @sunyi000,


If you run a `kubectl get cm -n abcdesktop`, you will find the config map.

But it's not the problem and you have found this issue.
You check that :
if I do a cat /etc/group, I see the last line being yisun:x:1102, which has the correct gid but wrong name.

I use [posixAccount](https://ldapwiki.com/wiki/PosixAccount), and only `uid` is defined.
The ldap query read this posixAccount attributs:

- [cn](https://ldapwiki.com/wiki/Cn)
- [uid](https://ldapwiki.com/wiki/Uid)
- [uidNumber](https://ldapwiki.com/wiki/UidNumber)
- [gidNumber](https://ldapwiki.com/wiki/GidNumber)
- [homeDirectory](https://ldapwiki.com/wiki/HomeDirectory)

There is no group name, or I misundestrand some thing. 
So to create the /etc/group `pyos` use the posixAccount `uid` as a group name ( and **this is wrong**, you wrote `which has the correct gid but wrong name`.)

Do you use the member attribut to define the group name of a user ?
My stupid question is : 

- How to find the value for a correct group name ?

Thank you so much.
Alexandre

**Comment by sunyi000** [comment]
thank you @alexandredevely 

sorry I stored my group schema in a separate place so I forgot to mention this.

for my ldap server, group name is in 'cn', and uniqueMember is the user in the group. 

I have cn, uniqueMember, description,ou in my group schema

I don't use memberOf attribute in users schema.

so for users dn will be  cn=yisun,cn=users,dc=example,dc=com
for group, it will be like cn=test_group,cn=groups,dc=example,dc=com

I just log on to the ldap server and do ldapsearch for the group id and cn

thanks

**Comment by alexandredevely** [comment]
thank you @sunyi000 
I undertstand why the groupe name is wrong.
I should be possible to support [PosixGroup](https://ldapwiki.com/wiki/PosixGroup) 
I need to install a openldap...

See you 
Alexandre

**Comment by sunyi000** [comment]
thank you..
Supporting posixGroup will be very helpful.

**Comment by sunyi000** [comment]
Hi @alexandredevely 
some more questions about apps..
- is the file manager app still in abcdesktopio/filemanager.d ? the latest tag is about a year ago..
- a naive question about debugging apps. in v1.1 i just use docker logs, and kubectl logs which seems enough for debugging apps. for 3.0 what is the best way to debug if my own app image doesn't work?

I deployed the filemanager.d a couple of my own apps, changing the base image to oc.template.gtk:3.0. I can see they all appear in the quick launcher bar, but none of them starts, so I'm just trying to debug. 

thanks very much

**Comment by sunyi000** [comment]
just to answer my own question 1. I found out that nautilus is the one to use :)

**Comment by alexandredevely** [comment]
Hello @sunyi000,

Yes the file mananger is abcdesktopio/nautilus.d:3.0

## To debug and create application 

You need to do the job of pyos.

### login on abcdesktop and start an application like nautilus for example

Start `nautilus` application for example 


The next command need a shell on your server 

### run `docker ps -a` to get the containerid of nautilus

```
docker ps -a
CONTAINER ID   IMAGE                       COMMAND                  CREATED              STATUS                          PORTS       NAMES
83ad0ce39b07   008d088e70a2                "/composer/appli-doc…"   7 seconds ago        Up 7 seconds                                k8s_philip-j--fry-nautilus-bca8fc830bfa48319f222da0fbed45f4_fry-6c601377-e70d-4e05-b241-978e9d501ed8_abcdesktop_b57301f4-ea1f-4c89-8d72-defb495551bb_0
```

### run `docker exec -it 83ad0ce39b07 sh -c 'echo $XAUTH_KEY'` to get the `XAUTH_KEY`

```
docker exec -it 83ad0ce39b07 sh -c 'echo $XAUTH_KEY'
82e2d940117c5f15e06e6148e1f481
```


### run `docker inspect  --format='{{.HostConfig.Binds}}'  83ad0ce39b07` to get the bind volume mapping

```
docker inspect --format='{{.HostConfig.Binds}}' 83ad0ce39b07
[/var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/tmp:/tmp /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/x11socket:/tmp/.X11-unix /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/pulseaudiosocket:/tmp/.pulseaudio /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/cupsdsocket:/tmp/.cupsd /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/run:/var/run/desktop /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/log:/var/log/desktop /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~secret/auth-localaccount-fry:/var/secrets/abcdesktop/localaccount:ro /mnt/philip-j--fry:/home/balloon /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/etc-hosts:/etc/hosts /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/containers/philip-j--fry-nautilus-bca8fc830bfa48319f222da0fbed45f4/6d31d007:/dev/termination-log]
```

Look for entries 
- `/tmp` 
- `/tmp/.X11-unix`
- `/home/balloon` 


### start an image `ubuntu:22.04` and bind it to the created volumes

I choose `ubuntu:22.04` but you can replace by you own.

```
docker run -v /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/tmp:/tmp -v  /var/lib/kubelet/pods/b57301f4-ea1f-4c89-8d72-defb495551bb/volumes/kubernetes.io~empty-dir/x11socket:/tmp/.X11-unix -v  /mnt/philip-j--fry:/home/balloon -it ubuntu:22.04 bash 
root@1ece60f84320:/# 
```

Take care you are `root` and not your local account.


### install xauth 

```
apt-get update
apt-get install xauth
export XAUTH_KEY=82e2d940117c5f15e06e6148e1f481
export DISPLAY=:0.0
xauth add $DISPLAY MIT-MAGIC-COOKIE-1 $XAUTH_KEY
```

### install a sample application

```
apt-get install x11-apps
```

You can start `xcalc` 

```
xcalc
```

**Comment by alexandredevely** [comment]
@sunyi000 

I need you to check the ldap account.
Can you confirm that this LDIF extract matchs your own LDAP account ?
If this is correct, i will commit this update.

I have created an account `name1` `uidnumber: 1000 gidnumber: 2000`
`name1` is member of `group1` and `greatgroup`

```
account uid=name1
dn: uid=name1,ou=Users,dc=example,dc=com
cn: name1
uid: name1
homeDirectory: /home/name1
loginShell: /bin/bash
description: fun
uidNumber: 1000
gidNumber: 2000
objectClass: posixAccount
objectClass: shadowAccount
```

A group for name1 with the same `gidnumber: 2000`

```
dn: cn=group1,ou=Groups,dc=example,dc=com
cn: group1
description: name1group
gidNumber: 2000
objectClass: posixGroup
```

> I dont't know if we need to add this attribut `memberUid: name1`(I think we don't, because it is already defined in name1)

And another group for `name1`

```
dn: cn=greatgroup,ou=Groups,dc=example,dc=com
cn: greatgroup
description: greatname1group
gidNumber: 3000
objectClass: posixGroup
memberUid: name1
memberUid: name3
memberUid: name2
```

This will make 

## /etc/passwd

```
name1:*:1000:2000:fun:/home/name1:/bin/bash
```

## /etc/group

```
group1:*:2000:
greatgroup:*:3000:name1,name2,name3
```

The `id` command returns

```
uid=1000(name1) gid=2000(group1) groups=2000(group1),3000(greatgroup)
```

Thank you so much

**Comment by alexandredevely** [comment]
Fix in release 3.0
